In [40]:
#import modules 

import matplotlib.pyplot as plt
import pandas as pd
import time
import numpy as np 

In [42]:
#read in our sample dataset
#comments can be seen from non-toxic to toxic in ascending order

df = pd.read_csv('Sample_labaled_data.csv')
df.sort_values(['toxic'])

Unnamed: 0                id  \
0               5  0001ea8717f6de06   
36177      100026  a6ee3e9d0cc68c1a   
36178      100027  a6eec4dc389e7aa6   
36179      100036  a6f3d3882ac1eb6d   
36180      100037  a6f48a298d74e56a   
...           ...               ...   
11961       33082  37004d98441d16aa   
29561       81915  88b1af0d03e94fa0   
3142         8780  0eb0723835202637   
45090      124622  d0327b2d72a39bfb   
33300       92146  9998b727ebd2790f   

                                            comment_text toxic  
0      Thank you for understanding I think very highl...    no  
36177    Watch this video  httpwwwyoutubecomwatchvhDZ...    no  
36178  References  This article needs to be better re...    no  
36179                       Thanks Ill see what I can do    no  
36180   ImageIndianapolisStateFairThisYear06jpg liste...    no  
...                                                  ...   ...  
11961                          The baddest b of them all   yes  
29561                            you is a fart face tehe   yes  
3142    Rajtarangani   I will over Rajtarangani and w...   yes  
45090  If you want to make yourself ridiculous please...   yes  
33300   1 February 2013 UTC  Oh that means putting th...   yes  

[55252 rows x 4 columns]

In [43]:
#transform strings to integers 
#toxic comments = 1, non-toxic labeled comments = 0 
df['toxic'] = [1 if x == 'yes' else 0 for x in df['toxic']]
df.to_csv('Sample_labaled_data_updated.csv', index=False)

#read in the updated dataset w toxic as ints
df_updated = pd.read_csv('Sample_labaled_data_updated.csv')
df_updated.sort_values(['toxic'])

Unnamed: 0                id  \
0               5  0001ea8717f6de06   
36177      100026  a6ee3e9d0cc68c1a   
36178      100027  a6eec4dc389e7aa6   
36179      100036  a6f3d3882ac1eb6d   
36180      100037  a6f48a298d74e56a   
...           ...               ...   
11961       33082  37004d98441d16aa   
29561       81915  88b1af0d03e94fa0   
3142         8780  0eb0723835202637   
45090      124622  d0327b2d72a39bfb   
33300       92146  9998b727ebd2790f   

                                            comment_text  toxic  
0      Thank you for understanding I think very highl...      0  
36177    Watch this video  httpwwwyoutubecomwatchvhDZ...      0  
36178  References  This article needs to be better re...      0  
36179                       Thanks Ill see what I can do      0  
36180   ImageIndianapolisStateFairThisYear06jpg liste...      0  
...                                                  ...    ...  
11961                          The baddest b of them all      1  
29561                            you is a fart face tehe      1  
3142    Rajtarangani   I will over Rajtarangani and w...      1  
45090  If you want to make yourself ridiculous please...      1  
33300   1 February 2013 UTC  Oh that means putting th...      1  

[55252 rows x 4 columns]

The from googleapiclient.discovery imports the build function to be used to utilize the Google Comment Analyzer API client, then being used to make requests to the API

In [14]:
from googleapiclient.discovery import build
import json

def get_toxicity_score(comment):
    API_KEY = 'AIzaSyCfmcFQAvQkcfNhpRA-9L7xWF3wLWLTG5k'
    
    client = build("commentanalyzer", "v1alpha", developerKey=API_KEY, discoveryServiceUrl = "https://commentanalyzer.googleapis.com/$discovery/rest?version=v1alpha1", static_discovery = False,)
    analyze_request = {'comment': { 'text': comment}, 'requestedAttributes': {'TOXICITY': {}}}
    response = client.comments().analyze(body=analyze_request).execute()
    toxicity_score = response["attributeScores"]["TOXICITY"]["summaryScore"]["value"]

    return toxicity_score

In [15]:
#take examples from the dataset to determine a threshold 
import numpy as np

# Generate some example data
scores = np.random.rand(100)
labels = np.where(scores > 0.5, 1, 0)

# Choose a threshold
threshold = 0.05

# Evaluate the performance of the threshold
tp = np.sum(np.logical_and(scores >= threshold, labels == 1))
fp = np.sum(np.logical_and(scores >= threshold, labels == 0))
tn = np.sum(np.logical_and(scores < threshold, labels == 0))
fn = np.sum(np.logical_and(scores < threshold, labels == 1))
precision = tp / (tp + fp)
recall = tp / (tp + fn)
f1_score = 2 * (precision * recall) / (precision + recall)

print("Threshold:", threshold)
print("Precision:", precision)
print("Recall:", recall)
print("F1-score:", f1_score)

Threshold: 0.05
Precision: 0.5520833333333334
Recall: 1.0
F1-score: 0.7114093959731543


the F1-score of 0.7114093959731543 indicates that the model does a well enough job between achieveing precision and recall, however there still do exist missclassifications of the comments. Overall, the score suggest the model performs somewhat well with plenty of room for imporvement, moreso in the area of precision that is marked by 0.5520833333333334

I hypothesize that this model will fail if we replace the most common swear words with less common obscenities. the sample example comments below are being pulled from the sample labaled dataset on Canvas. 

In [16]:
#develop a small test of N# of sample example comments to test this hypothesis 

#create sample set 
sample_comments = [
    "Simple You are stupid",
    "what the hell are you doing",
    "Hey shithead stop vandilizing articles",
    "where I feel like a man instead of an idiot"
    "If ya not still fuk u",
    "Karl Tearle is a mop haired twat"]

#create list of the lesser common obscentities to replace the common swear words
obscenities = ["douchebag",
              "jackass",
              "numbskull",
              "bonehead",
              "knucklehead",
              "dumbass"]


In [17]:
#create a function to replace the common swear wors with the obscenities
import random
def replace_obscenities(comment):
    words = comment.split()
    for i in range(len(words)):
        if words[i].lower() in ["stupid", "hell", "shithead", "idiot", "fuk", "twat"]:
            words[i] = random.choice(obscenities)
    return " ".join(words)

#generate a test of n# of samples to see
num_samples = 6
for i in range(num_samples):
    original_comment = random.choice(sample_comments)
    modified_comment = replace_obscenities(original_comment)
    #print og and modified comments
    print("Original Comment: ", original_comment)
    print("Modified Comment: ", modified_comment)

Original Comment:  what the hell are you doing
Modified Comment:  what the knucklehead are you doing
Original Comment:  Hey shithead stop vandilizing articles
Modified Comment:  Hey knucklehead stop vandilizing articles
Original Comment:  Karl Tearle is a mop haired twat
Modified Comment:  Karl Tearle is a mop haired numbskull
Original Comment:  where I feel like a man instead of an idiotIf ya not still fuk u
Modified Comment:  where I feel like a man instead of an idiotIf ya not still jackass u
Original Comment:  Simple You are stupid
Modified Comment:  Simple You are dumbass
Original Comment:  Karl Tearle is a mop haired twat
Modified Comment:  Karl Tearle is a mop haired knucklehead


In [19]:
#document model scores and assess hypothesis
for comment in sample_comments:
    score = get_toxicity_score(comment)
    print(comment, score)

HttpError: <HttpError 403 when requesting https://commentanalyzer.googleapis.com/$discovery/rest?version=v1alpha1&key=AIzaSyCfmcFQAvQkcfNhpRA-9L7xWF3wLWLTG5k returned "Perspective Comment Analyzer API has not been used in project 824922415462 before or it is disabled. Enable it by visiting https://console.developers.google.com/apis/api/commentanalyzer.googleapis.com/overview?project=824922415462 then retry. If you enabled this API recently, wait a few minutes for the action to propagate to our systems and retry.". Details: "[{'@type': 'type.googleapis.com/google.rpc.Help', 'links': [{'description': 'Google developers console API activation', 'url': 'https://console.developers.google.com/apis/api/commentanalyzer.googleapis.com/overview?project=824922415462'}]}, {'@type': 'type.googleapis.com/google.rpc.ErrorInfo', 'reason': 'SERVICE_DISABLED', 'domain': 'googleapis.com', 'metadata': {'service': 'commentanalyzer.googleapis.com', 'consumer': 'projects/824922415462'}}]">

Potential impact of a low sample size:

We know that larger sample sizes give us more accurate representation of the dataset at hand. Thus, a low sample size may be niche to a handful of comments, say if they are all toxic, then the toxicity score will be high. Whereas on the other hand, if the smaller sample size contains all non-toxic comments, the toxicity score will be low and isn't representative of the actual dataset that contains around 5000 rows of text. 